File: procedural/queens.erl
-module(queen).
-export([main/1]).

main([Arg]) ->
    Size = list_to_integer(atom_to_list(Arg)),
    Row = array:new(Size, {default, true}),
    Up = array:new(2 * Size - 1, {default, true}),
    Down = array:new(2 * Size - 1, {default, true}),
    Board = array:new(Size, {default, 0}),
    queen(0, Size, Row, Up, Down, Board),
    init:stop().

queen(Size, Size, _, _, _, Board) ->
    for(0, Size - 2, fun(I) -> io:format("~p ", [array:get(I, Board)]) end),
    io:format("~p~n", [array:get(Size - 1, Board)]);

queen(N, Size, Row, Up, Down, Board) ->
    for(0, Size - 1,
        fun(I) ->
                case array:get(I, Row)
                    and array:get(N+I, Up)
                    and array:get(N-I+Size-1, Down) of
                    true ->
                        queen(N + 1,
                            Size,
                            array:set(I, false, Row),
                            array:set(N+I, false, Up),
                            array:set(N-I+Size-1, false, Down),
                            array:set(N, I+1, Board));
                    false ->
                        none
                end
        end).

for(M, M, F) -> [F(M)];
for(I, M, F) -> [F(I) | for(I+1, M, F)].

File: procedural/queens.ml
open Array;;

let () =
    let size = int_of_string Sys.argv.(1) in
    let row = make size true
    and up = make (2 * size - 1) true
    and down = make (2 * size - 1) true
    and board = make size 0 in
    let rec queen n =
        if n >= size then begin
            for i = 0 to size - 2 do begin
                print_int board.(i); print_char ' '
            end done;
            print_int board.(size - 1); print_char '\n'
        end else
            for i = 0 to size - 1 do
                if row.(i) && up.(n+i) && down.(n-i+size-1) then begin
                    row.(i)<-false; up.(n+i)<-false; down.(n-i+size-1)<-false;
                    board.(n)<-i + 1;
                    queen(n + 1);
                    row.(i)<-true; up.(n+i)<-true; down.(n-i+size-1)<-true
                end
            done in
    queen 0;;
